### DISCUSSION TITLE
Bayer ordered dithering with custom palette

### DISCUSSION THREAD

**Discussion by clogged** [ImageMagick/ImageMagick#6619] [Help]
### ImageMagick version

ImageMagick 6.9.11-60 Q16 x86_64 2021-01-25

### Operating system, version and so on

Google Colab

### Description

Hi, 
I just want to start off by saying that I have no coding background and I've been doing all this by mostly reading stuff on various forums and using chatgpt.

I played around with this [repo](https://github.com/cj-holmes/ggboy) and I really liked the results. (making "pixelart" from regular images with bayer dither and custom palette.) I wanted to do some changes to it (add more then 4 colours to the palette, different resolutions, etc.), but couldn't figure out how to do it...

Anyways, I decided to create my own program with the help of chatgpt. Since the original repo used imagemagick - that's what I focused on.

Here is my process:

Original image:
![test](https://github.com/ImageMagick/ImageMagick/assets/80620913/44be3c26-a84c-4634-8809-962aaea78e5b)

```
!convert -resize x128 test.png resized.png

!convert resized.png -colorspace Gray gray.png

!convert gray.png -ordered-dither o4x4,8 dithered.png

!convert xc:"#23372e" xc:"#2c4c39" xc:"#466a36" xc:"#6b7312" +append colortable.gif

!convert dithered.png -dither None -remap colortable.gif remapped.png
```

Result:
![result](https://github.com/ImageMagick/ImageMagick/assets/80620913/af123f2d-6b1c-4546-9378-c6119bd13ad5)

Result from ggboy repo:
![result1](https://github.com/ImageMagick/ImageMagick/assets/80620913/1d72a605-b0de-4cad-951b-9f6663d89731)

I don't understand what am I doing wrong? For some reason my process is not remapping the colours as it "should". I've been stuck with this for weeks and I simply can't figure it out...
 
It probably has to something with this [part](https://github.com/cj-holmes/ggboy/blob/137370f0cedfa78c6ed5853b6f27ff31a8f5060d/R/ggboy.R) of the code from the ggboy repo, but I simply don't have the necessary knowledge to understand what exactly happens at this part:

```
 dplyr::mutate(original = (col2rgb(col) %>% t())[,1], # pixel intensities
                  original =  (original - min(original))/(max(original) - min(original)),
                  dither_mat = b_filter,
                  dithered = cut(original + (f*dither_mat/4), breaks = 4, labels = FALSE)) %>%
    dplyr::transmute(x = x + 16, y = y + 16, dithered = dithered) %>%
```

Would really appreciate it if someone could point me in the right direction.

Thank you!

**Comment by snibgo** [comment]
IM's `-remap` chooses, for each pixel, the closest colour from the map. "Closest" means the smallest Euclidean distance, in the current colorspace. The quoted code includes "`col2rgb(col)`", which suggests the colours may be translated to linear RGB. Something like this (Windows BAT syntax):
```
magick ^
  cartoonLips.png ^
  -resize x128 ^
  -colorspace Gray ^
  -set colorspace sRGB -colorspace RGB ^
  -ordered-dither o4x4,8 ^
  +write d.png ^
  ( xc:"#23372e" xc:"#2c4c39" xc:"#466a36" xc:"#6b7312" +append +repage ^
    -set colorspace sRGB -colorspace RGB ^
    +write mpr:COLTAB +delete ^
  ) ^
  -dither None -remap mpr:COLTAB ^
  -set colorspace RGB -colorspace sRGB ^
  x.png
```

![x](https://github.com/ImageMagick/ImageMagick/assets/33812368/cc978eb2-737b-4872-947c-fd4cccc830db)

EDIT: `magick` is for ImageMagick v7. I suggest using that, not the very old v6 `convert`.

**Reply by clogged** [reply]
This looks much better! But it still looks kinda overexposed if you compare it to the ggboy output... When I asked chatgp to explain that code snippet - this was the response:

![Screenshot 2023-09-06 at 3 13 34 PM](https://github.com/ImageMagick/ImageMagick/assets/80620913/97cd9d30-52a2-4313-9992-8357be6bd9dc)

It seems like only red channel values are exported and then dither is applied to it - if I understood that correctly?

One more thing I forgot to mention is that he also had this line after turning the image into grayscale  `magick::image_raster() %>%`, not sure if that might play the role in this?

Here is the full snippet btw:

```
p <-
    i %>%
    magick::image_scale(geometry = "128x112^") %>%
    magick::image_crop(geometry = "128x112", gravity = gravity) %>%
    magick::image_convert(type = "grayscale") %>%
    magick::image_raster() %>%
    dplyr::arrange(x) %>%
    dplyr::mutate(original = (col2rgb(col) %>% t())[,1], # pixel intensities
                  original =  (original - min(original))/(max(original) - min(original)),
                  dither_mat = b_filter,
                  dithered = cut(original + (f*dither_mat/4), breaks = 4, labels = FALSE)) %>%
    dplyr::transmute(x = x + 16, y = y + 16, dithered = dithered) %>%
    dplyr::bind_rows(frame_df %>% dplyr::mutate(dithered = c(1, 4)[frame]))

  if(length(palette) == 1 && (palette %in% 1:4)){
    palette <- palettes[[palette]]
  }
```
Rest of the explanation by chtagpt:

![Screenshot 2023-09-06 at 3 29 10 PM](https://github.com/ImageMagick/ImageMagick/assets/80620913/46b8185e-5e83-4994-bc5a-7a7d687519ad)

It also uses custom bayer matrix, but I don't think that's the reason.

Also, I'm using 6.9 version cause thats what I get when I run this command in google colab: `!apt-get install imagemagick`

Will look into how to get the version 7.

Thank you for your response, much appreciated!

**Comment by snibgo** [comment]
To get just the red channel:
```
-channel Red -separate +channel
```

> original =  (original - min(original))/(max(original) - min(original))

In IM, this can be done with:
```
-auto-level
```

I don't know what "`magick::image_raster()`" does. Perhaps it rearranges the pixel data in a format suitable for "`dplyr::arrange(x)`".

**Reply by clogged** [reply]
Unfortunately I'm still not getting the desired outputs... I think my question might be more appropriate to some R experts considering only few operations were done using imagemagick.

Once again, thank you for your help!
